File: sdk/python/tests/e2e_flow.py
"""E2E: FlowProducer — trees, chains, fan-in, getFlow."""

from __future__ import annotations

import threading

from harness import Server, test, unique_name, wait_until

from bunqueue import FlowProducer, Queue, Worker


@test
def flow_tree_children_before_parent(server: Server) -> None:
    qname = unique_name("flow")
    order = []
    lock = threading.Lock()

    def process(job):
        with lock:
            order.append(job.name)
        return {"step": job.name}

    with Queue(qname, port=server.port) as queue, FlowProducer(port=server.port) as flows:
        worker = Worker(qname, process, port=server.port, concurrency=4, poll_timeout_ms=300)
        worker.start()
        try:
            node = flows.add(
                {
                    "name": "parent",
                    "queueName": qname,
                    "data": {"kind": "root"},
                    "children": [
                        {"name": "child-a", "queueName": qname, "data": {}},
                        {"name": "child-b", "queueName": qname, "data": {}},
                    ],
                }
            )
            assert len(node.children) == 2
            assert wait_until(lambda: queue.get_state(node.job.id) == "completed", 20)
        finally:
            worker.close(timeout=10)

        assert order[-1] == "parent", f"parent did not run last: {order}"
        assert set(order[:2]) == {"child-a", "child-b"}
        values = queue.get_children_values(node.job.id)
        assert len(values) == 2, f"children values missing: {values}"


@test
def flow_chain_sequential(server: Server) -> None:
    qname = unique_name("chain")
    order = []
    lock = threading.Lock()

    def process(job):
        with lock:
            order.append(job.name)
        return job.name

    with Queue(qname, port=server.port) as queue, FlowProducer(port=server.port) as flows:
        worker = Worker(qname, process, port=server.port, concurrency=4, poll_timeout_ms=300)
        worker.start()
        try:
            ids = flows.add_chain(
                [
                    {"name": "step-1", "queueName": qname},
                    {"name": "step-2", "queueName": qname},
                    {"name": "step-3", "queueName": qname},
                ]
            )
            assert len(ids) == 3
            assert wait_until(lambda: queue.get_state(ids[-1]) == "completed", 20)
        finally:
            worker.close(timeout=10)
        assert order == ["step-1", "step-2", "step-3"], f"chain order broken: {order}"


@test
def flow_fan_in(server: Server) -> None:
    qname = unique_name("fanin")
    merged = {}

    def process(job):
        if job.name == "merge":
            merged["children"] = job.get_children_values()
            return {"merged": True}
        return {"part": job.name}

    with Queue(qname, port=server.port) as queue, FlowProducer(port=server.port) as flows:
        worker = Worker(qname, process, port=server.port, concurrency=4, poll_timeout_ms=300)
        worker.start()
        try:
            result = flows.add_bulk_then(
                [
                    {"name": "part-a", "queueName": qname},
                    {"name": "part-b", "queueName": qname},
                    {"name": "part-c", "queueName": qname},
                ],
                {"name": "merge", "queueName": qname},
            )
            assert len(result["parallel_ids"]) == 3
            assert wait_until(lambda: queue.get_state(result["final_id"]) == "completed", 20)
        finally:
            worker.close(timeout=10)
        assert len(merged.get("children", {})) == 3, f"fan-in values: {merged}"


@test
def get_flow_reconstructs_tree(server: Server) -> None:
    qname = unique_name("gettree")
    with Queue(qname, port=server.port) as queue, FlowProducer(port=server.port) as flows:
        queue.pause()  # keep jobs in place while we inspect
        node = flows.add(
            {
                "name": "root",
                "queueName": qname,
                "children": [
                    {
                        "name": "mid",
                        "queueName": qname,
                        "children": [{"name": "leaf", "queueName": qname}],
                    }
                ],
            }
        )
        tree = flows.get_flow(node.job.id)
        assert tree is not None
        assert len(tree.children) == 1
        assert len(tree.children[0].children) == 1
        queue.obliterate()


@test
def flow_custom_ids_and_reciprocal_links(server: Server) -> None:
    qname = unique_name("customflow")
    root_id = f"root-{qname}"
    child_id = f"child-{qname}"
    with Queue(qname, port=server.port) as queue, FlowProducer(port=server.port) as flows:
        queue.pause()
        node = flows.add(
            {
                "name": "root",
                "queueName": qname,
                "opts": {"job_id": root_id},
                "children": [
                    {
                        "name": "child",
                        "queueName": qname,
                        "opts": {"job_id": child_id},
                    }
                ],
            }
        )
        assert node.job.id == root_id
        assert node.job.custom_id == root_id
        assert node.children[0].job.id == child_id
        assert node.children[0].job.parent_id == root_id
        assert node.job.children_ids == [child_id]
        queue.obliterate()


@test
def flow_invalid_batch_commits_no_jobs(server: Server) -> None:
    qname = unique_name("rollback")
    with Queue(qname, port=server.port) as queue, FlowProducer(port=server.port) as flows:
        try:
            flows.add_chain(
                [
                    {"name": "ok-step", "queueName": qname},
                    {"name": "bad-step", "queueName": "invalid queue name!!"},
                ]
            )
            raise AssertionError("expected chain to fail on invalid queue name")
        except Exception:
            pass
        assert queue.count() == 0, "failed atomic batch created jobs"

File: sdk/typescript/tests/e2e-flow.ts
/** E2E: FlowProducer — atomic trees, chains, fan-in and flow reads. */

import { FlowProducer, type Job } from '../dist/legacy.js';
import {
  assert,
  assertEq,
  getPort,
  makeWorker,
  namedQueue,
  qname,
  test,
  waitFor,
} from './harness.ts';

function makeFlow(): FlowProducer {
  return new FlowProducer({ host: '127.0.0.1', port: getPort() });
}

test('flow: tree children complete before parent + getChildrenValues', async () => {
  const name = qname('flowtree');
  const queue = namedQueue(name);
  const flow = makeFlow();
  const order: string[] = [];
  const worker = makeWorker(
    name,
    async (job) => {
      order.push(job.name ?? '?');
      return { step: job.name };
    },
    { concurrency: 4 }
  );
  try {
    const node = await flow.add({
      name: 'parent',
      queueName: name,
      data: { kind: 'root' },
      children: [
        { name: 'child-a', queueName: name, data: {} },
        { name: 'child-b', queueName: name, data: {} },
      ],
    });
    assertEq(node.children?.length, 2, 'two children in node');
    await waitFor(async () => (await queue.getJobState(node.job.id)) === 'completed', 20_000);
    assertEq(order[order.length - 1], 'parent', `parent ran last: ${order.join(',')}`);
    assertEq(new Set(order.slice(0, 2)).size, 2, 'children ran first');
    const values = await queue.getChildrenValues(node.job.id);
    assertEq(Object.keys(values).length, 2, 'children values');
  } finally {
    await worker.close();
    flow.close();
    queue.close();
  }
});

test('flow: addChain executes sequentially', async () => {
  const name = qname('chain');
  const queue = namedQueue(name);
  const flow = makeFlow();
  const order: string[] = [];
  const worker = makeWorker(
    name,
    async (job) => {
      order.push(job.name ?? '?');
      return job.name;
    },
    { concurrency: 4 }
  );
  try {
    const { jobIds } = await flow.addChain([
      { name: 'step-1', queueName: name },
      { name: 'step-2', queueName: name },
      { name: 'step-3', queueName: name },
    ]);
    assertEq(jobIds.length, 3, 'three chained jobs');
    await waitFor(async () => (await queue.getJobState(jobIds[2])) === 'completed', 20_000);
    assertEq(order.join(','), 'step-1,step-2,step-3', 'chain order');
  } finally {
    await worker.close();
    flow.close();
    queue.close();
  }
});

test('flow: addBulkThen fan-in reads children values', async () => {
  const name = qname('fanin');
  const queue = namedQueue(name);
  const flow = makeFlow();
  let childrenSeen: Record<string, unknown> | null = null;
  const worker = makeWorker(
    name,
    async (job: Job) => {
      if (job.name === 'merge') {
        childrenSeen = await job.getChildrenValues();
        return { merged: true };
      }
      return { part: job.name };
    },
    { concurrency: 4 }
  );
  try {
    const { parallelIds, finalId } = await flow.addBulkThen(
      [
        { name: 'part-a', queueName: name },
        { name: 'part-b', queueName: name },
        { name: 'part-c', queueName: name },
      ],
      { name: 'merge', queueName: name }
    );
    assertEq(parallelIds.length, 3, 'three parallel jobs');
    await waitFor(async () => (await queue.getJobState(finalId)) === 'completed', 20_000);
    assert(childrenSeen !== null, 'merge read children values');
    assertEq(Object.keys(childrenSeen ?? {}).length, 3, 'three children values');
  } finally {
    await worker.close();
    flow.close();
    queue.close();
  }
});

test('flow: getFlow reconstructs nested tree', async () => {
  const name = qname('gettree');
  const queue = namedQueue(name);
  const flow = makeFlow();
  try {
    await queue.pause(); // keep jobs in place while we inspect
    const node = await flow.add({
      name: 'root',
      queueName: name,
      children: [
        {
          name: 'mid',
          queueName: name,
          children: [{ name: 'leaf', queueName: name }],
        },
      ],
    });
    const tree = await flow.getFlow({ id: node.job.id });
    assert(tree !== null, 'tree fetched');
    assertEq(tree?.children?.length, 1, 'one mid child');
    assertEq(tree?.children?.[0].children?.length, 1, 'one leaf child');
    await queue.obliterate();
  } finally {
    flow.close();
    queue.close();
  }
});

test('flow: custom IDs and reciprocal links survive the atomic snapshot', async () => {
  const name = qname('customflow');
  const rootId = `root-${name}`;
  const childId = `child-${name}`;
  const queue = namedQueue(name);
  const flow = makeFlow();
  try {
    await queue.pause();
    const node = await flow.add({
      name: 'root',
      queueName: name,
      opts: { jobId: rootId },
      children: [{ name: 'child', queueName: name, opts: { jobId: childId } }],
    });
    assertEq(node.job.id, rootId, 'root planned custom ID');
    assertEq(node.job.customId, rootId, 'root customId snapshot');
    assertEq(node.children?.[0].job.id, childId, 'child planned custom ID');
    assertEq(node.children?.[0].job.parentId, rootId, 'child points to root');
    assertEq(node.job.childrenIds[0], childId, 'root owns child');
    await queue.obliterate();
  } finally {
    flow.close();
    queue.close();
  }
});

test('flow: invalid batch commits no jobs', async () => {
  const name = qname('rollback');
  const queue = namedQueue(name);
  const flow = makeFlow();
  try {
    let threw = false;
    try {
      await flow.addChain([
        { name: 'ok-step', queueName: name },
        { name: 'bad-step', queueName: 'invalid queue name!!' },
      ]);
    } catch {
      threw = true;
    }
    assert(threw, 'chain with invalid queue must throw');
    assertEq(await queue.count(), 0, 'failed atomic batch created no jobs');
  } finally {
    flow.close();
    queue.close();
  }
});
